api

```diff
@@ -351,6 +351,55 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
   echo "pi-apps-$(echo "$app" | md5sum | cut -c1-8 | awk '{print $1}')"
 }
 
+add_external_repo() { # add an external apt repo and its gpg key
+  repopath="$1"
+  reponame="$2"
+  pubkeyurl="$3"
+
+  # check if all needed vars are set
+  [ -z "$repopath" ] && error "add_external_packages: repopath not set"
+  [ -z "$reponame" ] && error "add_external_packages: reponame not set"
+  [ -z "$pubkeyurl" ] && error "add_external_packages: pubkeyurl not set"
+
+  # check if links are valid
+  wget -q --spider "$pubkeyurl" || error "add_external_packages: pubkeyurl isn't a valid link"
+
+  # make pi-apps key storage dir if it doesn't exist
+  if [ ! -d /usr/share/keyrings ]; then
+    sudo mkdir -p /usr/share/keyrings || error "add_external_repo: failed to create pi-apps key storage dir."
+  fi
+
+  # check if .list file already exists
+  [ -f /etc/apt/sources.list.d/${reponame}.list ] && error "add_external_repo: /etc/apt/sources.list.d/${reponame}.list already exists."
+
+  # download gpg key from specified url
+  rm -f /usr/share/keyrings/${reponame}-keyring.gpg
+  curl -fsSL "$pubkeyurl" | sudo gpg --dearmor -o /usr/share/keyrings${reponame}-keyring.gpg
+
+  if [ $? != 0 ];then
+    sudo rm -f /etc/apt/sources.list.d/${reponame}.list
+    sudo rm -f /usr/share/keyrings/${reponame}-keyring.gpg
+    error "add_external_repo: download from specified pubkeyurl failed."
+  fi
+
+  # create .list file
+  echo "deb [signed-by=/usr/share/keyrings/$reponame-keyring.gpg] $repopath" | sudo tee /etc/apt/sources.list.d/${reponame}.list >/dev/null
+
+  if [ $? != 0 ];then
+    sudo rm -f /etc/apt/sources.list.d/${reponame}.list
+    sudo rm -f /usr/share/keyrings/${reponame}-keyring.gpg
+    error "add_external_repo: failed to create .list file"
+  fi
+}
+
+rm_external_repo() { # remove an external apt repo and its gpg key
+  reponame="$1"
+
+  rm -f /usr/share/keyrings/${reponame}-keyring.gpg || error "rm_external_repo: removal of ${reponame}-keyring.gpg failed"
+
+  sudo rm -f /etc/apt/sources.list.d/${reponame}.list || error "rm_external_repo: removal of ${reponame}.list failed"
+}
```

Comment:
I still think this function should be made to be much more like "remove_repofile_if_unused" (thats an existing function)

we shouldn't always remove a repo because it may be in use in multiple apps.

Reply:
makes sense. I will make some changes later tomorrow, it's getting late here.